tests/components/kodi/__init__.py

```diff
@@ -11,13 +11,24 @@
     CONF_USERNAME,
 )
 
-from .util import MockConnection
+from .util import MockConnection, MockWSConnection
 
 from tests.common import MockConfigEntry
 
+PATCH_KODI = "homeassistant.components.kodi"
```

Comment:
You should patch the library code, not the HA integration code

Reply:
In the `init_integration` function I use `"pykodi.kodi..."` whenever possible, but in the case of `get_kodi_connection`it is not working. I figured that's because `get_kodi_connection` is imported as a function on file level and then called in the context of `...kodi_connman.` instead of `pykodi.kodi.`.

`f"{PATCH_KODI}.kodi_connman.get_kodi_connection", return_value=conn(_connected)` only this returned the correct connection for the test.

Reply:
Can you try patching "homeassistant.components.kodi.kodi_connman.get_kodi_connection"?
Also, that file should probably be renamed to `connection_manager.py`

Reply:
> Can you try patching "homeassistant.components.kodi.kodi_connman.get_kodi_connection"?

I did that. Your suggestion = `f"{PATCH_KODI}.kodi_connman.get_kodi_connection", return_value=conn(_connected)`

>  renamed to connection_manager.py

I can do that.

Reply:
right, but everything else that's patched is integration code. The next 3 lines are all things that should be replaced with patching the library.

Reply:
Actually I have some good reasons for those:
* `f"{PATCH_KODI_MEDIA}.KodiEntity.async_update"`
  this I use because I don't want to handle all the player related stuff here for the binary sensor tests and stop the errors with it.
* `f"{PATCH_KODI_BINARY}._reset_state"`
  this is the only way to validate that the exception was caught in the application code. I just assert that it has to be called.
* `f"{PATCH_KODI_CONNMAN}.register_websocket_callback"`
  that's the last function before the jsonrpc library. Where `setattr` saves the callback functions. In the jsonrpc_base this gets to `__setattr__` which then calls `"jsonrpc_base.jsonrpc.Server.__register"`. I suppose because the mock connection returns `None` for the property `server` I cannot get there?!

Reply:
I also tried `f"tests.components.kodi.MockWSConnection.server", new_callable=PropertyMock()` instead. But this does not get called. The [docu (last lines of that chapter)](https://docs.python.org/3/library/unittest.mock.html#mocking-magic-methods) suggests that `__setattr__` is not supported.

Reply:
Ok, I got rid of those, now using only the sensor states.